File: packages/compute/src/production.ts
/**
 * Production rate integration and Bateman equations.
 *
 * Computes energy-integrated production rates, time-dependent activity,
 * daughter ingrowth, saturation yield, and depth profiles.
 */

import { LN2, MILLIBARN_CM2, ELEMENTARY_CHARGE, MEV_TO_JOULE } from "./constants";
import { interp, linspace, trapezoid } from "./interpolation";

/**
 * Minimum peak cross-section (mb) on the integration grid to consider
 * the reaction channel active.  Values below this are treated as zero
 * to prevent interpolation artifacts from generating spurious
 * production rates.
 */
const MIN_PEAK_XS_MB = 1e-6;

/**
 * Compute energy-integrated production rate for one isotope.
 *
 * R = beam_particles/s * (n_atoms / V) * integral(sigma(E) / |dE/dx(E)| dE)
 */
export function computeProductionRate(
  xsEnergiesMeV: Float64Array,
  xsMb: Float64Array,
  dedxFn: (energy: Float64Array) => Float64Array,
  energyInMeV: number,
  energyOutMeV: number,
  nTargetAtoms: number,
  beamParticlesPerS: number,
  targetVolumeCm3: number,
  nPoints: number = 100,
): {
  productionRate: number;
  energies: Float64Array;
  xsInterp: Float64Array;
  dedxValues: Float64Array;
} {
  // Clamp lower bound above zero: log-log stopping-power interpolation
  // produces NaN at E = 0 (log(0) = -Infinity).
  const energies = linspace(Math.max(energyOutMeV, 0.01), energyInMeV, nPoints);

  // Interpolate cross-section onto grid (zero outside data range)
  const xsInterp = interp(energies, xsEnergiesMeV, xsMb, 0, 0);

  // Evaluate stopping power
  const dedxValues = dedxFn(energies);

  // Skip channels where the interpolated XS is effectively zero across the
  // entire energy range.  This prevents interpolation artifacts from
  // producing non-zero production rates for reactions that have no
  // appreciable cross-section at the operating energies.
  let peakXs = 0;
  for (let i = 0; i < nPoints; i++) {
    if (xsInterp[i] > peakXs) peakXs = xsInterp[i];
  }
  if (peakXs < MIN_PEAK_XS_MB) {
    return { productionRate: 0, energies, xsInterp, dedxValues };
  }

  // Integrand: sigma(E) / |dE/dx(E)|
  const integrand = new Float64Array(nPoints);
  for (let i = 0; i < nPoints; i++) {
    integrand[i] = xsInterp[i] / Math.abs(dedxValues[i]);
  }

  const integral = trapezoid(integrand, energies);

  const numberDensity = nTargetAtoms / targetVolumeCm3;
  const prate = beamParticlesPerS * numberDensity * integral * MILLIBARN_CM2;

  return { productionRate: prate, energies, xsInterp, dedxValues };
}

/**
 * Compute time-dependent activity using Bateman equations.
 *
 * During irradiation: A(t) = R * (1 - exp(-λt))
 * During cooling: A(t) = A(T_irr) * exp(-λ(t - T_irr))
 */
export function batemanActivity(
  productionRate: number,
  halfLifeS: number | null,
  irradiationTimeS: number,
  coolingTimeS: number,
  nTimePoints: number = 200,
): { timeGrid: Float64Array; activity: Float64Array } {
  const nIrr = Math.floor(nTimePoints / 2);
  const nCool = nTimePoints - nIrr;

  const tIrr = linspace(0, irradiationTimeS, nIrr);
  const tCoolFull = linspace(irradiationTimeS, irradiationTimeS + coolingTimeS, nCool + 1);
  // Exclude duplicate point at irradiationTimeS
  const tCool = tCoolFull.slice(1);

  const timeGrid = new Float64Array(nIrr + nCool);
  timeGrid.set(tIrr, 0);
  timeGrid.set(tCool, nIrr);

  const activity = new Float64Array(timeGrid.length);

  if (halfLifeS === null || halfLifeS <= 0) {
    return { timeGrid, activity };
  }

  const lambda = LN2 / halfLifeS;

  // Irradiation phase
  for (let i = 0; i < timeGrid.length; i++) {
    if (timeGrid[i] <= irradiationTimeS) {
      activity[i] = productionRate * (1 - Math.exp(-lambda * timeGrid[i]));
    }
  }

  // Activity at end of irradiation
  const aEoi = productionRate * (1 - Math.exp(-lambda * irradiationTimeS));

  // Cooling phase
  for (let i = 0; i < timeGrid.length; i++) {
    if (timeGrid[i] > irradiationTimeS) {
      const dtCool = timeGrid[i] - irradiationTimeS;
      activity[i] = aEoi * Math.exp(-lambda * dtCool);
    }
  }

  return { timeGrid, activity };
}

/**
 * Compute daughter activity from parent decay during cooling (Bateman ingrowth).
 */
export function daughterIngrowth(
  parentActivityEoiBq: number,
  parentHalfLifeS: number,
  daughterHalfLifeS: number | null,
  branchingRatio: number,
  coolingTimesS: Float64Array,
): Float64Array {
  const result = new Float64Array(coolingTimesS.length);

  if (daughterHalfLifeS === null || daughterHalfLifeS <= 0) {
    return result;
  }

  const lambdaP = LN2 / parentHalfLifeS;
  const lambdaD = LN2 / daughterHalfLifeS;

  // Degenerate case: lambda_P ≈ lambda_D
  if (Math.abs(lambdaD - lambdaP) < 1e-30 * Math.max(lambdaD, lambdaP)) {
    for (let i = 0; i < result.length; i++) {
      const t = coolingTimesS[i];
      result[i] =
        branchingRatio * parentActivityEoiBq * lambdaD * t * Math.exp(-lambdaD * t);
    }
    return result;
  }

  const coeff =
    (branchingRatio * parentActivityEoiBq * lambdaD) / (lambdaD - lambdaP);
  for (let i = 0; i < result.length; i++) {
    const t = coolingTimesS[i];
    result[i] = coeff * (Math.exp(-lambdaP * t) - Math.exp(-lambdaD * t));
  }
  return result;
}

/** Compute saturation yield [Bq/µA]. */
export function saturationYield(
  productionRate: number,
  halfLifeS: number | null,
  beamCurrentMA: number,
): number {
  if (halfLifeS === null || halfLifeS <= 0) return 0;
  const currentUA = beamCurrentMA * 1e3;
  return productionRate / currentUA;
}

/**
 * Generate depth profile from integration points.
 * Returns cumulative depths and volumetric heat deposition.
 */
export function generateDepthProfile(
  energies: Float64Array,
  dedxValues: Float64Array,
  beamCurrentMA: number,
  areaCm2: number,
  projectileZ: number,
): { depths: Float64Array; energiesOrdered: Float64Array; heatWCm3: Float64Array } {
  const n = energies.length;

  // Input arrives as energyOut→energyIn (low→high). Reverse so index 0 = beam
  // entry (highest energy) and depth accumulates toward energyOut.
  const eRev = new Float64Array(n);
  const dRev = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    eRev[i] = energies[n - 1 - i];
    dRev[i] = dedxValues[n - 1 - i];
  }

  const depths = new Float64Array(n);
  const dE = n > 1 ? Math.abs(eRev[0] - eRev[1]) : 0;

  for (let i = 1; i < n; i++) {
    depths[i] = depths[i - 1] + dE / Math.abs(dRev[i - 1]);
  }

  const beamParticlesPerS =
    (beamCurrentMA * 1e-3) / (projectileZ * ELEMENTARY_CHARGE);
  const flux = beamParticlesPerS / areaCm2;

  const heatWCm3 = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    heatWCm3[i] = flux * Math.abs(dRev[i]) * MEV_TO_JOULE;
  }

  return { depths, energiesOrdered: eRev, heatWCm3 };
}

/**
 * Compute local production rate [atoms/s/cm] at each depth point
 * for a single reaction channel.
 *
 * rate(x) = flux × n_density × σ(E(x)) × weight × MILLIBARN_CM2
 */
export function computeDepthProductionRate(
  xsEnergiesMeV: Float64Array,
  xsMb: Float64Array,
  depthEnergiesMeV: Float64Array,
  numberDensity: number,
  beamParticlesPerS: number,
  areaCm2: number,
  weight: number,
): Float64Array {
  const n = depthEnergiesMeV.length;
  const flux = beamParticlesPerS / areaCm2;
  const xsAtDepth = interp(depthEnergiesMeV, xsEnergiesMeV, xsMb, 0, 0);
  const result = new Float64Array(n);
  const scale = flux * numberDensity * weight * MILLIBARN_CM2;
  for (let i = 0; i < n; i++) {
    result[i] = scale * xsAtDepth[i];
  }
  return result;
}

File: core/src/production.rs
//! Production rate integration and depth profiles.

use crate::constants::{ELEMENTARY_CHARGE, MEV_TO_JOULE, MILLIBARN_CM2, MIN_PEAK_XS_MB};
use crate::interpolation::{interp, linspace, trapezoid};

/// Result of production rate computation.
pub struct ProductionRateResult {
    pub production_rate: f64,
    pub energies: Vec<f64>,
    pub xs_interp: Vec<f64>,
    pub dedx_values: Vec<f64>,
}

/// Compute energy-integrated production rate for one isotope.
///
/// R = beam_particles/s * (n_atoms / V) * integral(sigma(E) / |dE/dx(E)| dE)
pub fn compute_production_rate(
    xs_energies_mev: &[f64],
    xs_mb: &[f64],
    dedx_fn: &dyn Fn(&[f64]) -> Vec<f64>,
    energy_in_mev: f64,
    energy_out_mev: f64,
    n_target_atoms: f64,
    beam_particles_per_s: f64,
    target_volume_cm3: f64,
    n_points: usize,
) -> ProductionRateResult {
    let e_low = energy_out_mev.max(0.01);
    let energies = linspace(e_low, energy_in_mev, n_points);

    // Interpolate cross-section onto grid (zero outside data range)
    let xs_interp = interp(&energies, xs_energies_mev, xs_mb, 0.0, 0.0);

    // Evaluate stopping power
    let dedx_values = dedx_fn(&energies);

    // Check if XS is effectively zero
    let peak_xs = xs_interp.iter().cloned().fold(0.0_f64, f64::max);
    if peak_xs < MIN_PEAK_XS_MB {
        return ProductionRateResult {
            production_rate: 0.0,
            energies,
            xs_interp,
            dedx_values,
        };
    }

    // Integrand: sigma(E) / |dE/dx(E)|
    let integrand: Vec<f64> = xs_interp
        .iter()
        .zip(dedx_values.iter())
        .map(|(&xs, &dedx)| xs / dedx.abs())
        .collect();

    let integral = trapezoid(&integrand, &energies);
    let number_density = n_target_atoms / target_volume_cm3;
    let prate = beam_particles_per_s * number_density * integral * MILLIBARN_CM2;

    ProductionRateResult {
        production_rate: prate,
        energies,
        xs_interp,
        dedx_values,
    }
}

/// Saturation yield [Bq/µA].
pub fn saturation_yield(
    production_rate: f64,
    half_life_s: Option<f64>,
    beam_current_ma: f64,
) -> f64 {
    match half_life_s {
        Some(t) if t > 0.0 => {
            let current_ua = beam_current_ma * 1e3;
            production_rate / current_ua
        }
        _ => 0.0,
    }
}

/// Depth profile result.
pub struct DepthProfileResult {
    pub depths: Vec<f64>,
    pub energies_ordered: Vec<f64>,
    pub heat_w_cm3: Vec<f64>,
}

/// Generate depth profile from integration points.
/// Input energies arrive as energyOut→energyIn (low→high).
pub fn generate_depth_profile(
    energies: &[f64],
    dedx_values: &[f64],
    beam_current_ma: f64,
    area_cm2: f64,
    projectile_z: u32,
) -> DepthProfileResult {
    let n = energies.len();

    // Reverse so index 0 = beam entry (highest energy)
    let e_rev: Vec<f64> = energies.iter().rev().copied().collect();
    let d_rev: Vec<f64> = dedx_values.iter().rev().copied().collect();

    let mut depths = vec![0.0; n];
    let de = if n > 1 {
        (e_rev[0] - e_rev[1]).abs()
    } else {
        0.0
    };

    for i in 1..n {
        depths[i] = depths[i - 1] + de / d_rev[i - 1].abs();
    }

    let beam_particles_per_s = (beam_current_ma * 1e-3) / (projectile_z as f64 * ELEMENTARY_CHARGE);
    let flux = beam_particles_per_s / area_cm2;

    let heat_w_cm3: Vec<f64> = d_rev
        .iter()
        .map(|&d| flux * d.abs() * MEV_TO_JOULE)
        .collect();

    DepthProfileResult {
        depths,
        energies_ordered: e_rev,
        heat_w_cm3,
    }
}

/// Local production rate [atoms/s/cm] at each depth point.
pub fn compute_depth_production_rate(
    xs_energies_mev: &[f64],
    xs_mb: &[f64],
    depth_energies_mev: &[f64],
    number_density: f64,
    beam_particles_per_s: f64,
    area_cm2: f64,
    weight: f64,
) -> Vec<f64> {
    let flux = beam_particles_per_s / area_cm2;
    let xs_at_depth = interp(depth_energies_mev, xs_energies_mev, xs_mb, 0.0, 0.0);
    let scale = flux * number_density * weight * MILLIBARN_CM2;
    xs_at_depth.iter().map(|&xs| scale * xs).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Drift guard: the energy-integrated production rate
    /// (compute_production_rate) and the spatial integral of the depth-resolved
    /// rate (compute_depth_production_rate) are the same physical quantity
    /// expressed in different coordinates. They must agree within numerical
    /// integration tolerance for any XS / dEdx / energy window.
    ///
    /// If this test ever fails, one of the two formulas drifted — fix the one
    /// that was changed, not the tolerance.
    #[test]
    fn depth_rate_integrates_to_production_rate() {
        // Flat σ = 50 mb in [5, 100] MeV; zero outside.
        let xs_energies_mev: Vec<f64> = (5..=100).map(|e| e as f64).collect();
        let xs_mb: Vec<f64> = xs_energies_mev.iter().map(|_| 50.0).collect();

        // Roughly proton-on-Al stopping power, varying with E to avoid trivial cases.
        let dedx_fn = |energies: &[f64]| -> Vec<f64> {
            energies.iter().map(|&e| 40.0 / e.sqrt()).collect()
        };

        let energy_in = 72.0;
        let energy_out = 8.0;
        let particles_per_s = 2.5e14; // ~40 µA of protons
        let area_cm2 = 1.0;
        let density_g_cm3 = 2.70; // Al
        let avg_a = 26.98;
        let number_density = density_g_cm3 * crate::constants::AVOGADRO / avg_a;
        let weight = 0.5; // non-trivial weight to exercise the multiplication

        // For compute_production_rate we hand it n_atoms / volume so they agree.
        let thickness_cm = 0.8; // arbitrary for this test
        let volume_cm3 = thickness_cm * area_cm2;
        let n_target_atoms = number_density * volume_cm3 * weight;

        let integrated = compute_production_rate(
            &xs_energies_mev,
            &xs_mb,
            &dedx_fn,
            energy_in,
            energy_out,
            n_target_atoms,
            particles_per_s,
            volume_cm3,
            400,
        );

        // Depth profile shares the same energies + dedx values the integrator used.
        let depth = generate_depth_profile(
            &integrated.energies,
            &integrated.dedx_values,
            /* current_ma, unused for depth-rate comparison */ 1.0,
            area_cm2,
            /* projectile_z */ 1,
        );

        let depth_rates = compute_depth_production_rate(
            &xs_energies_mev,
            &xs_mb,
            &depth.energies_ordered,
            number_density,
            particles_per_s,
            area_cm2,
            weight,
        );

        let integrated_from_depth =
            crate::interpolation::trapezoid(&depth_rates, &depth.depths);

        let rel_err =
            ((integrated_from_depth - integrated.production_rate) / integrated.production_rate).abs();
        assert!(
            rel_err < 0.005,
            "drift: ∫(depth_rate) dx = {:.6e}, compute_production_rate = {:.6e}, rel_err = {:.4}",
            integrated_from_depth,
            integrated.production_rate,
            rel_err
        );
    }

    /// Foreign-layer property: a target isotope with weight = 0 (e.g. Cu-63
    /// inside a pure Al layer) must produce zero rate at every depth point.
    /// This guarantees the stack-plot's "goes to zero in foreign layers"
    /// behaviour falls out of the physics, without special boundary logic.
    #[test]
    fn foreign_layer_rate_is_identically_zero() {
        let xs_energies_mev: Vec<f64> = (5..=100).map(|e| e as f64).collect();
        let xs_mb: Vec<f64> = xs_energies_mev.iter().map(|_| 123.4).collect();
        let depth_energies = vec![72.0, 50.0, 25.0, 10.0];

        let rates = compute_depth_production_rate(
            &xs_energies_mev,
            &xs_mb,
            &depth_energies,
            /* number_density */ 6.02e22,
            /* particles_per_s */ 1e14,
            /* area_cm2 */ 1.0,
            /* weight */ 0.0,
        );

        assert_eq!(rates.len(), 4);
        assert!(rates.iter().all(|&r| r == 0.0), "got {:?}", rates);
    }

    /// Below-threshold property: an XS table that starts at 30 MeV gives zero
    /// rate at depths where E(x) has dropped below 30 MeV. Verifies the
    /// interp(left=0, right=0) threshold handling still holds.
    #[test]
    fn below_threshold_rate_is_zero() {
        let xs_energies_mev: Vec<f64> = (30..=100).map(|e| e as f64).collect();
        let xs_mb: Vec<f64> = xs_energies_mev.iter().map(|_| 100.0).collect();
        let depth_energies = vec![72.0, 50.0, 40.0, 32.0, 20.0, 10.0, 1.0];

        let rates = compute_depth_production_rate(
            &xs_energies_mev,
            &xs_mb,
            &depth_energies,
            6.02e22,
            1e14,
            1.0,
            1.0,
        );

        // Above threshold: non-zero. Below threshold (depth_energies[4..]): zero.
        assert!(rates[0] > 0.0);
        assert!(rates[1] > 0.0);
        assert!(rates[2] > 0.0);
        assert!(rates[3] > 0.0);
        assert_eq!(rates[4], 0.0);
        assert_eq!(rates[5], 0.0);
        assert_eq!(rates[6], 0.0);
    }
}
